File: UnionFind/UnionFind.cpp
#include <vector>
#include <unordered_map>

class UnionFind {
private:
    std::unordered_map<int, int> parent;
    std::unordered_map<int, int> rank;
public:
    UnionFind(const std::vector<int>& a);

    int find(int x);
    void merge(int x, int y);

    bool contains(int x);
};

UnionFind::UnionFind(const std::vector<int>& a) {
    for (int x : a) {
        parent[x] = x;
        rank[x] = 0;
    }
}

int UnionFind::find(int x) {
    if (parent[x] != x) {
        parent[x] = find(parent[x]);
    }
    return parent[x];
}

void UnionFind::merge(int x, int y) {
    x = find(x);
    y = find(y);

    if (x == y) {
        return;
    }

    if (rank[x] < rank[y]) {
        std::swap(x, y);
    }
    parent[y] = x;
    rank[x] = std::max(rank[x], rank[y] + 1);
}

bool UnionFind::contains(int x) {
    return parent.find(x) != parent.end();
}

File: Union Find/UnionFind.cs
using System;
using System.Collections.Generic;

public class UnionFind
{
    private readonly Dictionary<int, int> _parent = new();
    private readonly Dictionary<int, int> _rank = new();

    public UnionFind(IReadOnlyCollection<int> a)
    {
        foreach (var x in a)
        {
            _parent[x] = x;
            _rank[x] = 0;
        }
    }

    public int Find(int x)
    {
        if (_parent[x] != x)
            _parent[x] = Find(_parent[x]);

        return _parent[x];
    }

    public void Merge(int x, int y)
    {
        x = Find(x);
        y = Find(y);

        if (x == y)
            return;

        if (_rank[x] < _rank[y])
            (x, y) = (y, x);

        _parent[y] = x;
        _rank[x] = Math.Max(_rank[x], _rank[y] + 1);
    }

    public bool Contains(int x) => _parent.ContainsKey(x);
}
